library/threads.lisp

```diff
@@ -0,0 +1,322 @@
+    "Release `lock'. It is an error to call this unless
+the lock has previously been acquired (and not released) by the same
+thread. If other threads are waiting for the lock, the
+`acquire-lock' call in one of them will now be able to continue.
+
+Returns the lock."
+    (lisp (Result LispCondition RecursiveLock) (lock)
+      (cl:handler-case (Ok (bt2:release-recursive-lock lock))
+        (cl:error (c) (Err c)))))
+
+  (declare with-recursive-lock-held (RecursiveLock -> (Unit -> :a) -> :a))
+  (define (with-recursive-lock-held lock thunk)
+    (acquire-recursive-lock lock)
+    (let ((result (thunk)))
+      (release-recursive-lock lock)
+      result)))
+
+;;---------------------;;
+;; Condition Variables ;;
+;;---------------------;;
+
+(coalton-toplevel
+  (repr :native bt2:condition-variable)
+  (define-type ConditionVariable)
+
+  (declare make-cv (Unit -> ConditionVariable))
+  (define (make-cv)
+    "Creates a condition variable."
+    (lisp ConditionVariable ()
+      (bt2:make-condition-variable)))
+
+  (declare await-cv (ConditionVariable -> Lock -> Unit))
+  (define (await-cv cv lock)
+    "Atomically release `lock' and enqueue the calling thread waiting for `cv'.
+The thread will resume when another thread has notified it using `notify-cv';
+it may also resume if interrupted by some external event or in other
+implementation-dependent circumstances: the caller must always test on waking
+that there is threading to be done, instead of assuming that it can go ahead."
+    (lisp Unit (cv lock)
+      (bt2:condition-wait cv lock)
+      Unit))
+
+  (declare notify-cv (ConditionVariable -> Unit))
+  (define (notify-cv cv)
+    "Notify one of the threads waiting for `cv'."
+    (lisp Unit (cv)
+      (bt2:condition-notify cv)
+      Unit))
+
+  (declare broadcast-cv (ConditionVariable -> Unit))
+  (define (broadcast-cv cv)
+    "Notify all of the threads waiting for `cv'."
+    (lisp Unit (cv)
+      (bt2:condition-broadcast cv)
+      Unit)))
+
+;;------------;;
+;; Semaphores ;;
+;;------------;;
+
+(coalton-toplevel
+  (repr :native bt2:semaphore)
+  (define-type Semaphore)
+
+  (declare make-semaphore (Unit -> Semaphore))
+  (define (make-semaphore)
+    "Creates a semaphore with initial count 0."
+    (lisp Semaphore ()
+      (bt2:make-semaphore)))
+
+  (declare signal-semaphore (Semaphore -> UFix -> Unit))
+  (define (signal-semaphore sem count)
+    "Increment `sem' by `count'.
+If there are threads awaiting this semaphore, then `count' of them are woken up."
+    (lisp Unit (sem count)
+      (bt2:signal-semaphore sem :count count)
+      Unit))
+
+  (declare await-semaphore (Semaphore -> Unit))
+  (define (await-semaphore sem)
+    "Decrement the count of `sem' by 1 if the count is larger than zero.
+If the count is zero, blocks until `sem' can be decremented."
+    (lisp Unit (sem)
+      (bt2:wait-on-semaphore sem)
+      Unit)))
+
+;;---------;;
+;; Atomics ;;
+;;---------;;
+
+(coalton-toplevel
```

Comment:
I'm not sure the API below should have conditionals in it. Maybe atomic values shouldn't be included in the thread API for now.